### PR TITLE
[Analysis Epic] (4) Analysis products

### PR DESCRIPTION
**Pull request by profjsb**
This is the fourth PR in the Analysis epic. It's the last backend/middleware-focused PR, and should be merged after #3134 (many of the putative changes currently shown are carried over from #3134 and disappear upon merge). The remaining PRs will be frontend focused.

This PR adds an API endpoint that allows programmatic access to the products produced by an analysis service. In particular we add endpoints for: plots, results, and posterior visualization. These will be called by the front-end to show the results to the end users.

The GET endpoints are:

Plots (return an image to display):
```
http://localhost:5000/api/obj/analysis/<analysis_id>/plots/<plot_number> 
```

Posterior viz (ie. return a corner plot image to display):
```
http://localhost:5000/api/obj/analysis/<analysis_id>/corner
```
(attach corner plot kwargs in the body of the GET)

Results (returned as a json):
```
http://localhost:5000/api/obj/analysis/<analysis_id>/results
```

For example, to save the first plot of analysis_id=32, one can do something like this:

```python
import shutil
url = "http://localhost:5000/api/obj/analysis/32/plots/0"
r = requests.get(url, headers=header, stream=True)
if r.status_code == 200:
    with open('img.png', 'wb') as out_file:
        shutil.copyfileobj(r.raw, out_file)
```

### CODE REVIEW

File: doc/analysis.md [hunk 1]
```diff
@@ -112,8 +112,39 @@ results = joblib.load(base64.b64decode(data["analysis"]["results"]["data"]))
 
 A valid POST will immediately invalidate the unique token so that analysis entry cannot be posted to again (the user can simply restart an analysis with different parameters if they wish).
 
+### Getting specific analysis products programmatically
 
-### Supernova Fitter Example
+SkyPortal has API endpoint that allows programmatic access to GET products (plots, results, and posterior visualization) produced by an analysis service.
```
Comment (mcoughlin) [inline doc/analysis.md:117]: an API endpoint

File: skyportal/models/analysis.py [hunk 2]
```diff
@@ -189,17 +199,148 @@ def authinfo(self, value):
...
+            temp_range = [
+                [
+                    inference_data["posterior"][x].data.min(),
+                    inference_data["posterior"][x].data.max(),
+                    x,
+                ]
+                for x in inference_data["posterior"]
+            ]
+            for x in temp_range:
+                if x[0] == x[1]:
```
Comment (mcoughlin) [inline skyportal/models/analysis.py:313]: What exactly happens here could use a comment

**Review by mcoughlin — APPROVED** [review]
@profjsb Very little to say really, looking forward to seeing how it works in practice.

File: .github/workflows/test_models.yaml [hunk 3]
```diff
@@ -139,7 +139,7 @@ jobs:
           fi
           sudo tar -xzf ${CACHED_DOWNLOAD_DIR}/geckodriver-v${GECKO_VER}-linux64.tar.gz -C /usr/local/bin
           geckodriver --version
-          pip install --upgrade selenium
+          pip install selenium==4.2.0
```
Comment (guynir42) [inline .github/workflows/test_models.yaml:142]: Do we still need to pin this version?

File: skyportal/handlers/api/analysis.py [hunk 4]
```diff
@@ -964,15 +1019,167 @@ def delete(self, analysis_resource_type, analysis_id):
...
+                if analysis.data in [None, {}]:
+                    return self.error("No data found for this Analysis.", status=404)
+
+                if product_type.lower() == "corner":
+                    if not analysis.has_inference_data:
+                        return self.error(
+                            "No inference data found for this Analysis.", status=404
+                        )
+
+                    plot_kwargs = self.get_query_argument("plot_kwargs", {})
```
Comment (guynir42) [inline skyportal/handlers/api/analysis.py:1132]: Aren't these supposed to be body arguments (according to the docs)?

File: skyportal/models/analysis.py [hunk 5]
```diff
@@ -189,17 +199,148 @@ def authinfo(self, value):
...
+    def generate_corner_plot(self, **plot_kwargs):
+        """Generate a corner plot of the posterior from the inference data."""
+
+        if not self.has_inference_data:
+            return None
+
+        # we could add different formats here in the future
+        # but for now we only support netcdf4 formats
+        if self.data['inference_data']["format"] != "netcdf4":
+            return None
```
Comment (guynir42) [inline skyportal/models/analysis.py:288]: Shouldn't this be an error of some sort?

File: skyportal/tests/api/test_analysis.py [hunk 6]
```diff
@@ -489,6 +495,16 @@ def test_run_analysis_with_correct_and_incorrect_token(
             False
         ), f"analysis was not started properly ({data['data']['status_message']})"
 
+    # Since this is random data, this fit might (usually) or fail (seldom)
```
Comment (guynir42) [inline skyportal/tests/api/test_analysis.py:498]:
```suggestion
    # Since this is random data, this fit might succeed (usually) or fail (seldom)
```

**Review by guynir42 — APPROVED** [review]
Minor comments. Looks good overall.

File: skyportal/handlers/api/analysis.py [hunk 7]
```diff
@@ -964,15 +1019,167 @@ def delete(self, analysis_resource_type, analysis_id):
...
             return self.success()
         else:
             return self.error(
                 f'analysis_resource_type must be one of {", ".join(["obj"])}',
                 status=404,
             )
+
+
+def serialize_results_data():
+    pass
```
[the remark this answers is not in the document]
Comment (profjsb) [reply skyportal/handlers/api/analysis.py:1041]: Extraneous -- will remove.

File: skyportal/models/analysis.py (hunk 5 above)
Comment (profjsb) [reply skyportal/models/analysis.py:288]: Yes...and now I catch it in the API handler.

**Comment by profjsb** [comment]
unrelated GHA test failures

**Pull request merged**